<h1 align="center">Assignment 1</h1>
<h2 align="center">Programming Machine Learning Applications [DSC 478]</h2>
<h3 align="center">Jonesh Shrestha (2222011)</h3>

In [148]:
# import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#### 1. **[5 pts]** Download the data set adult-modified-09-13-2025.csv and load it into an appropriate data structure such as a Pandas dataframe. Explore the general characteristics of the data as a whole: examine the means, standard deviations, and other statistics associated with the numeric attributes and frequencies associated with categorical attributes.

In [149]:
adult_census_df = pd.read_csv('adult-modified-09-13-2025.csv')
adult_census_df.head(10)

age      workclass  education  education-num         marital-status  \
0  39         Public  Bachelors             13          Never-married   
1  50  Self-employed  Bachelors             13                Married   
2  38        Private    HS-grad              9               Divorced   
3  53        Private       11th              7                Married   
4  28        Private  Bachelors             13                Married   
5  37        Private    Masters             14                Married   
6  49        Private        9th              5  Married-spouse-absent   
7  52  Self-employed    HS-grad              9                Married   
8  31        Private    Masters             14          Never-married   
9  42        Private  Bachelors             13                Married   

          occupation   relationship   race     sex  hours-per-week  \
0       Adm-clerical  Not-in-family  White    Male              40   
1    Exec-managerial        Husband  White    Male              13   
2  Handlers-cleaners  Not-in-family  White    Male              40   
3  Handlers-cleaners        Husband  Black    Male              40   
4     Prof-specialty           Wife  Black  Female              40   
5    Exec-managerial           Wife  White  Female              40   
6      Other-service  Not-in-family  Black  Female              16   
7    Exec-managerial        Husband  White    Male              45   
8     Prof-specialty  Not-in-family  White  Female              50   
9    Exec-managerial        Husband  White    Male              40   

  native-country income  
0  United-States  <=50K  
1  United-States  <=50K  
2  United-States  <=50K  
3  United-States  <=50K  
4           Cuba  <=50K  
5  United-States  <=50K  
6        Jamaica  <=50K  
7  United-States   >50K  
8  United-States   >50K  
9  United-States   >50K

In [150]:
adult_census_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  object
 1   workclass       32561 non-null  object
 2   education       32561 non-null  object
 3   education-num   32561 non-null  int64 
 4   marital-status  32561 non-null  object
 5   occupation      32561 non-null  object
 6   relationship    32561 non-null  object
 7   race            32561 non-null  object
 8   sex             32561 non-null  object
 9   hours-per-week  32561 non-null  int64 
 10  native-country  32561 non-null  object
 11  income          32561 non-null  object
dtypes: int64(2), object(10)
memory usage: 3.0+ MB


In [151]:
adult_census_df.groupby("education")["education-num"].unique()

education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education-num, dtype: object

In [152]:
adult_census_df.describe(include="all")

age workclass education  education-num marital-status  \
count   32561     32561     32561   32561.000000          32561   
unique     74         6        16            NaN              6   
top        36   Private   HS-grad            NaN        Married   
freq      897     22696     10501            NaN          14999   
mean      NaN       NaN       NaN      10.080679            NaN   
std       NaN       NaN       NaN       2.572720            NaN   
min       NaN       NaN       NaN       1.000000            NaN   
25%       NaN       NaN       NaN       9.000000            NaN   
50%       NaN       NaN       NaN      10.000000            NaN   
75%       NaN       NaN       NaN      12.000000            NaN   
max       NaN       NaN       NaN      16.000000            NaN   

            occupation relationship   race    sex  hours-per-week  \
count            32561        32561  32561  32561    32561.000000   
unique              15            6      5      2             NaN   
top     Prof-specialty      Husband  White   Male             NaN   
freq              4140        13193  27816  21790             NaN   
mean               NaN          NaN    NaN    NaN       40.437456   
std                NaN          NaN    NaN    NaN       12.347429   
min                NaN          NaN    NaN    NaN        1.000000   
25%                NaN          NaN    NaN    NaN       40.000000   
50%                NaN          NaN    NaN    NaN       40.000000   
75%                NaN          NaN    NaN    NaN       45.000000   
max                NaN          NaN    NaN    NaN       99.000000   

       native-country income  
count           32561  32561  
unique             42      2  
top     United-States  <=50K  
freq            29170  24720  
mean              NaN    NaN  
std               NaN    NaN  
min               NaN    NaN  
25%               NaN    NaN  
50%               NaN    NaN  
75%               NaN    NaN  
max               NaN    NaN

*First I loaded the csv dataset as a dataframe using pandas libarary. I used head(10) function to see the first 10 rows of the dataframe to check it is loaded correctly. Then I used info() to see the number of non-null count and the datatype of each column. It showed that there were no null values and education-num, hours-per-week has type int64, all other columns was of object datatype. I used describe(include="all") to check the means, standard deviations and other statistics for the numeric attributes. For education-num, the mean education-num was around 10, which suggests that on average people had 10th grade education, minimum education people had was Preschool and maximum education was 16 i.e. Doctorate. The standard deviation 2.57 shows that education-num is on average 2.57 unit far from the mean*

#### 2. **[10 pts]** Using Pandas, compute the number of missing values for each attribute in the data. Fill the missing values for all numeric attributes using the mean value for the attribute. After filling in the missing numeric values, drop all rows where a categorical attribute contains a missing value. Next, remove columns education and native-country. Show that the final resulting dataframe does not contain missing values and display the new mean and standard deviations for the numeric attributes. For the remaining parts of this assignment, you should use this new dataframe without missing values and the aforementioned columns.

In [153]:
adult_census_df.isna().sum()

age               0
workclass         0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [154]:
for col in adult_census_df.columns:
    print(col)
    print(adult_census_df[col].unique())

age
['39' '50' '38' '53' '28' '37' '49' '52' '31' '42' '30' '23' '32' '40'
 '34' '25' '43' '54' '35' '?' '59' '56' '19' '20' '45' '22' '48' '21' '24'
 '57' '44' '41' '29' '18' '47' '46' '36' '79' '27' '67' '76' '17' '55'
 '61' '70' '64' '33' '71' '68' '66' '51' '58' '26' '60' '90' '75' '65'
 '77' '62' '63' '80' '72' '74' '69' '73' '81' '78' '88' '82' '83' '84'
 '85' '86' '87']
workclass
['Public' 'Self-employed' 'Private' '?' 'Without-pay' 'Never-worked']
education
['Bachelors' 'HS-grad' '11th' 'Masters' '9th' 'Some-college' 'Assoc-acdm'
 'Assoc-voc' '7th-8th' 'Doctorate' 'Prof-school' '5th-6th' '10th'
 '1st-4th' 'Preschool' '12th']
education-num
[13  9  7 14  5 10 12 11  4 16 15  3  6  2  1  8]
marital-status
['Never-married' 'Married' 'Divorced' 'Married-spouse-absent' 'Separated'
 'Widowed']
occupation
['Adm-clerical' 'Exec-managerial' 'Handlers-cleaners' 'Prof-specialty'
 'Other-service' 'Sales' 'Craft-repair' 'Transport-moving'
 'Farming-fishing' 'Machine-op-inspct' 'Tech-support'

*Looking at the unique values from each column, we can see that '?' is used to denote missing values.*

In [155]:
(adult_census_df == '?').sum()

age                 45
workclass         1836
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
hours-per-week       0
native-country     583
income               0
dtype: int64

In [156]:
adult_census_df.replace('?', np.nan, inplace=True)
adult_census_df.isna().sum()

age                 45
workclass         1836
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
hours-per-week       0
native-country     583
income               0
dtype: int64

In [157]:
# To replace the numerical attributes missing values using the mean, we first need to convert those columns to int
adult_census_df['age'] = adult_census_df['age'].astype(float)

In [158]:
age_mean = adult_census_df['age'].mean()
adult_census_df.fillna({'age': age_mean}, axis=0, inplace=True)

In [164]:
# drop all rows where a categorical attribute contains a missing value.
adult_census_df.drop(adult_census_df[adult_census_df.workclass.isnull()].index, axis=0, inplace=True)
adult_census_df.drop(adult_census_df[adult_census_df.occupation.isnull()].index, axis=0, inplace=True)
adult_census_df.drop(adult_census_df[adult_census_df['native-country'].isnull()].index, axis=0, inplace=True)

In [168]:
# remove columns education and native-country. 
adult_census_df.drop(columns=['education', 'native-country'], inplace=True)

In [171]:
adult_census_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30162 entries, 0 to 32560
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             30162 non-null  float64
 1   workclass       30162 non-null  object 
 2   education-num   30162 non-null  int64  
 3   marital-status  30162 non-null  object 
 4   occupation      30162 non-null  object 
 5   relationship    30162 non-null  object 
 6   race            30162 non-null  object 
 7   sex             30162 non-null  object 
 8   hours-per-week  30162 non-null  int64  
 9   income          30162 non-null  object 
dtypes: float64(1), int64(2), object(7)
memory usage: 2.5+ MB


In [172]:
adult_census_df.describe()

age  education-num  hours-per-week
count  30162.000000   30162.000000    30162.000000
mean      38.438107      10.121312       40.931238
std       13.128006       2.549995       11.979984
min       17.000000       1.000000        1.000000
25%       28.000000       9.000000       40.000000
50%       37.000000      10.000000       40.000000
75%       47.000000      13.000000       45.000000
max       90.000000      16.000000       99.000000

#### 3. **[10 pts]** For the three numeric attributes (age, hours-per-week, education-num), display box plots that show the overall dispersion and skew in these variables. Next, create histograms for these three variables showing the overall data distribution in each. Finally, display a scatter plot of age (x-axis) vs. education-num (y-axis).

In [ ]:
# display box plots that show the overall dispersion and skew
